File: 백준/Silver/11052. 카드 구매하기/카드 구매하기.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	var n int
	fmt.Fscanf(reader, "%d\n", &n)
	var numbers = make([]int, n+1)
	var results = make([]int, n+1)

	//var count int
	for i := 1; i <= n; i++ {
		fmt.Fscanf(reader, "%d ", &numbers[i])
	}
	if n == 1 {
		fmt.Fprintf(writer, "%d", numbers[1])
	} else {

		results[1] = numbers[1]
		for i := 2; i <= n; i++ {
			results[i] = numbers[i] //기존값을 넣어두고
			for j := 1; j < i; j++ {
				//result를 만들때까지의 모든 경우의 수에 대하여 검사해서 최댓값을 넣음
				//ex. 5를 만드는 방법은 1*5, 2+1*2, 3+2, 3+1*2 등 다양함
				results[i] = int(math.Max(float64(results[i]), float64(results[i-j]+numbers[j])))
			}
		}
		fmt.Fprintf(writer, "%d\n", results[n])
	}
}

File: 백준/Silver/11052. 카드 구매하기/카드 구매하기.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
n = int(input())
numbers = list(map(int, input().split()))
numbers.insert(0, 0) #계산 편하게 하려고 끼워넣음
results = [0]*(n+1)
results[1]=numbers[1]
if n==1:
    print(numbers[1])
else:
    for i in range(2, n+1):
        results[i]=numbers[i]
        for j in range(1, i):
            results[i]=max(results[i], results[i-j]+numbers[j])
    print(results[n])